doctor_visits/delphi_doctor_visits/sensorize.py

```diff
@@ -0,0 +1,290 @@
+# third party
+import numpy as np
+import pandas as pd
+
+# first party
+from .config import Config
+
+class Sensorizer:
+    """Sensorizer class to convert indicator into a sensor.
+    """
+
+    @staticmethod
+    def linear_regression_coefs(
+            grouped_df,
+            signal_col="signal",
+            target_col="target",
+            fit_intercept=True):
+        """
+        Calculate linear regression coefficients for a grouped df.
+        For each group, calculate slope as cov(x,y)/var(x) and
+        intercept as mean(y) - slope*mean(x).
+
+        Args:
+            grouped_df: Grouped DataFrame, calculation to be done
+                        for each group
+            signal_col: column name of covariate
+            target_col: column name of response
+            fit_intercept: boolean whether to fit intercept
+
+        Returns:
+            DataFrame with columns for groups, and
+            "b1": slope of regression, and
+            "b0": intercept of regression
+        """
+
+        signal_mean = signal_col + "_mean"
+        target_mean = target_col + "_mean"
+        signal_var = signal_col + "_var"
+
+        if fit_intercept:
+            cov_df = grouped_df.apply(
+                        lambda x: x[signal_col].cov(x[target_col],ddof=0))
+            cov_df = cov_df.reset_index().rename(columns={0: "cov"})
+
+            var_df = grouped_df.apply(np.var).reset_index()
+            var_df = var_df.rename(
+                columns={signal_col:signal_var}).drop(target_col,axis=1)
+            means_df = grouped_df.agg(np.mean).reset_index()
+            means_df = means_df.rename(
+                columns={signal_col: signal_mean, target_col:target_mean})
+
+            fit_df = cov_df.merge(var_df).merge(means_df)
+            fit_df["b1"] = fit_df["cov"] / fit_df[signal_var]
+            fit_df["b0"] = \
+                fit_df[target_mean]-fit_df["b1"]*fit_df[signal_mean]
+            fit_df = fit_df.drop(
+                ["cov",signal_mean,target_mean,signal_var],axis=1)
+        else:
+            # No intercept, b1 = sum(x_i * y_i)/sum(x_i^2)
+            fit_df = grouped_df.apply(lambda x:
+                np.sum(x[signal_col]*x[target_col])/np.sum(x[signal_col]**2))
+            fit_df = fit_df.reset_index().rename(columns={0: "b1"})
+            fit_df["b0"] = 0
+        return fit_df
+
+
+    @staticmethod
+    def weighted_moments(df, data_col, wt_col):
+        """
+        Calculate weighted mean and std for a DataFrame
+
+        Args:
+            df: pd.DataFrame
+            data_col: column name of data
+            wt_col: column name of weight
+
+        Returns:
+            Tuple: (weighted mean, weighted_std)
+        """
+
+        w_mean = (df[data_col] * df[wt_col]).sum() / df[wt_col].sum()
+        w_var = (np.power(df[data_col]-w_mean,2) * df[wt_col]).sum()
+        w_var = w_var / df[wt_col].sum()
+        return (w_mean, np.sqrt(w_var))
+
+
+    @staticmethod
+    def sensorize(
```

Comment:
There's a lot going on here -- I'd split it up into a few different methods that get called depending on the various conditions